File: java/src/main/java/cn/az/code/medium/RotateList061.java
package cn.az.code.medium;

import cn.az.code.common.ListNode;

/**
 * @author Liz
 * @date 1/7/2020
 */
public class RotateList061 {

    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) {
            return null;
        }
        ListNode copyHead = head;
        int len = 1;
        while (copyHead.next != null) {
            copyHead = copyHead.next;
            len++;
        }
        //link the tail and head to make it a circle
        copyHead.next = head;
        for (int i = len - k % len; i > 1; i--) {
            head = head.next;
        }
        copyHead = head.next;
        //break the circle
        head.next = null;
        return copyHead;
    }
}

File: python/medium/RotateList061.py
from common import ListNode


def rotate_list(head: ListNode, k: int) -> ListNode:
    n, pre, cur = 0, None, head
    while cur:
        pre, cur = cur, cur.next
        n += 1
    if not n or k % n:
        return head
    tail = head
    for _ in range(n - k % n - 1):
        tail = tail.next
    nxt, tail.next, pre.next = tail.next, None, head
    return nxt
